src/libraries/System.Private.CoreLib/src/System/IO/SyncWindowsFileStreamStrategy.cs

```diff
@@ -0,0 +1,169 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Win32.SafeHandles;
+
+namespace System.IO
+{
+    internal sealed class SyncWindowsFileStreamStrategy : WindowsFileStreamStrategy
+    {
+        internal SyncWindowsFileStreamStrategy(SafeFileHandle handle, FileAccess access) : base(handle, access)
+        {
+        }
+
+        internal SyncWindowsFileStreamStrategy(string path, FileMode mode, FileAccess access, FileShare share, FileOptions options)
+            : base(path, mode, access, share, options)
+        {
+        }
+
+        internal override bool IsAsync => false;
+
+        protected override void OnInitFromHandle(SafeFileHandle handle)
+        {
+            // As we can accurately check the handle type when we have access to NtQueryInformationFile we don't need to skip for
+            // any particular file handle type.
+
+            // If the handle was passed in without an explicit async setting, we already looked it up in GetDefaultIsAsync
+            if (!handle.IsAsync.HasValue)
+                return;
+
+            // If we can't check the handle, just assume it is ok.
+            if (!(FileStreamHelpers.IsHandleSynchronous(handle, ignoreInvalid: false) ?? true))
+                throw new ArgumentException(SR.Arg_HandleNotSync, nameof(handle));
```

Comment:
It would not be beneficial as I would have to cover two cases where the arg name is called "handle" and "_fileHandle". It's easy to do, but overall would add more complexity to `ThrowHelper` than it's worth it.

https://github.com/dotnet/runtime/blob/6ef4b2e7aba70c514d85c2b43eac1616216bea55/src/libraries/System.Private.CoreLib/src/System/IO/FileStreamHelpers.Windows.cs#L149

https://github.com/dotnet/runtime/blob/6ef4b2e7aba70c514d85c2b43eac1616216bea55/src/libraries/System.Private.CoreLib/src/System/IO/LegacyFileStreamStrategy.Windows.cs#L437

Reply:
What do you mean by adding more complecity to `ThrowHelper`?

Can't you just add a method in said class as the following?
```cs
internal static void ThrowArgumentException_HandleNotSync(string argName)
{
    throw new ArgumentException(SR.Arg_HandleNotSync, argName);
}